app/backend/src/couchers/servicers/pages.py

```diff
@@ -0,0 +1,126 @@
+import logging
+
+import grpc
+
+from couchers import errors
+from couchers.db import session_scope
+from couchers.models import Page, PageType, PageVersion, User
+from couchers.utils import Timestamp_from_datetime, create_coordinate, remove_duplicates_retain_order, slugify
+from pb import pages_pb2, pages_pb2_grpc
+
+
+def _check_update_permission(page: Page, user_id):
+    if page.owner_user_id == user_id:
+        return True
+    # elif page.owner_cluster_id # TODO
+    return False
+
+
+def _page_to_pb(page: Page, user_id):
+    first_version = page.versions[0]
+    current_version = page.versions[-1]
+    lat, lng = current_version.coordinates
+    return pages_pb2.Page(
+        page_id=page.id,
+        slug=slugify(current_version.title),
+        created=Timestamp_from_datetime(first_version.created),
+        last_edited=Timestamp_from_datetime(current_version.created),
+        last_editor_user_id=current_version.editor_user_id,
+        creator_user_id=page.creator_user_id,
+        owner_user_id=page.owner_user_id,
+        owner_cluster_id=page.owner_cluster_id,
+        thread_id=None,  # TODO
+        title=current_version.title,
+        content=current_version.content,
+        address=current_version.address,
+        location=pages_pb2.Coordinate(
+            lat=lat,
+            lng=lng,
+        ),
+        editor_user_ids=remove_duplicates_retain_order([version.editor_user_id for version in page.versions]),
+        can_edit=_check_update_permission(page, user_id),
+    )
+
+
+class Pages(pages_pb2_grpc.PagesServicer):
+    def CreatePage(self, request, context):
+        if not request.title:
+            context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.MISSING_PAGE_TITLE)
+        if not request.content:
+            context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.MISSING_PAGE_CONTENT)
+        if not request.address:
+            context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.MISSING_PAGE_ADDRESS)
+        if not request.location.lat and not request.location.lng:
+            context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.MISSING_PAGE_LOCATION)
+
+        with session_scope() as session:
+            page = Page(
+                type=PageType.point_of_interest,
+                creator_user_id=context.user_id,
+                owner_user_id=context.user_id,
+            )
+            session.add(page)
+            session.flush()
+            page_version = PageVersion(
+                page=page,
+                editor_user_id=context.user_id,
+                title=request.title,
+                content=request.content,
+                address=request.address,
+                geom=create_coordinate(request.location.lat, request.location.lng),
+            )
+            session.add(page_version)
+            session.commit()
```

Comment:
wouldn't it be nicer to not do a commit and return after the context manager. Beginners might think the context manager does not commit. I also think it better matches the Zen of Python (PEP 20) and expecially: There should be one-- and preferably only one --obvious way to do it.

Reply:
I think it's clearer if the commit is before the `_page_to_pb`, which might even need the page_version to have been committed.